devito/passes/iet/misc.py

```diff
@@ -104,9 +113,88 @@ def relax_incr_dimensions(iet, **kwargs):
+
+            # Collect assumptions based on hierarchy
+            acs = [j for j in defines if j is not i.dim and j.is_Incr]
+            assumptions = []
+            assumptions.extend(chain.from_iterable((Gt(j.step, k.step), Gt(j, k)) for j, k
+                               in product(acs, acs) if j._depth > k._depth))
+
+            # Step 3: Reduce defines min/max candidates using assumptions on `defines`
+            defmin = reduce_relation(max, defines, assumptions)
+            defmax = reduce_relation(min, defines, assumptions)
+
+            # Collect symbolic_min, symbolic_max from defmin/defmax lists
+            defmin = [j.symbolic_min for j in defmin]
+            defmax = [j.symbolic_max for j in defmax]
+
+            # Drop duplicates from defmin/defmax and keep them Ordered
+            defmin = list(OrderedDict.fromkeys(defmin))
+            defmax = list(OrderedDict.fromkeys(defmax))
+
+            # Step 4: Drop a candidate if it is the symbolic_min/max of symbol candidates
+            for k in defmin.copy():  # TOFIX: Should not use a copy I guess?
+                try:
+                    if k.symbolic_min in defmin:
+                        defmin.remove(k.symbolic_min)
+                except:
+                    pass
+
+            for k in defmax.copy():
+                try:
+                    if k.symbolic_max in defmax:
+                        defmax.remove(k.symbolic_max)
+                except:
+                    pass
+
+            # At this point the candidates for the start and end of the `Iteration`
+            # comprise of candidates stemming (i) from `i.dim` and (ii) candidates
+            # stemming from other dims, (in space blocking other dims is usually root.)
+            # The canidates stemming from 'i.dim' can be compared and for those we need
+            # the minimum of them for the start and the maximum of them for the end so
+            # as to span the whole extent of the Iteration's Dimension (i.dim iteration
+            # extent).
+            # We split on comparable and uncomparable. Comparable ones can be further
+            # simplfied, (subdimensions, offsets incurred) and non-comparable ones
+            # include candidates that cannot be simplified more (no clear rule of
+            # further simplification exists)
+            # For the start: MAX(uncomparable, MIN(comparable)) --> MAX(uncomparable)
+            # For the end: MIN(uncomparable, MAX(comparable)) --> MIN(uncomparable)
+            defmin = reduce_relation(min, defmin, assumptions=None)
+            defmax = reduce_relation(max, defmax, assumptions=None)
+
+            # Apply offset if needed, else leave as it is
+            defmin = [j.subs(min_map) if j in min_map else j for j in as_list(defmin)]
+            defmax = [j.subs(max_map) if j in max_map else j for j in as_list(defmax)]
+
+            # Final MIN/MAX processing
+            iter_min = evaluate_relation(max, defmin)
```

Comment:
so you eventually call `reduce_relation` 3 times for MIN (and 3 times for MAX).

Why do you need 3 times for MIN (MAX) and not just one? that's really really odd at first glance

3 because one is buried inside `evaluate_relation`

Reply:
The first one is applied with assumptions on defines (no min max).
Second one is applied without assumptions on min/max.
The last one is indeed redundant, just removed.